### PR TITLE
[RISCV] Merge shuffle sources if lanes are disjoint

### PR DESCRIPTION
**Pull request by lukel97**
In x264, there's a few kernels with shuffles like this:

    %41 = add nsw <16 x i32> %39, %40
    %42 = sub nsw <16 x i32> %39, %40
    %43 = shufflevector <16 x i32> %41, <16 x i32> %42, <16 x i32> <i32 11, i32 15, i32 7, i32 3, i32 26, i32 30, i32 22, i32 18, i32 9, i32 13, i32 5, i32 1, i32 24, i32 28, i32 20, i32 16>

Because this is a complex two-source shuffle, this will get lowered as two vrgather.vvs that are blended together.

    vadd.vv v20, v16, v12
    vsub.vv v12, v16, v12
    vrgatherei16.vv v24, v20, v10
    vrgatherei16.vv v24, v12, v16, v0.t

However the indices coming from each source are disjoint, so we can blend the two together and perform a single source shuffle instead:

    %41 = add nsw <16 x i32> %39, %40
    %42 = sub nsw <16 x i32> %39, %40
    %43 = select <0,0,0,0,1,1,1,1,0,0,0,0,1,1,1,1> %41, %42
    %44 = shufflevector <16 x i32> %43, <16 x i32> poison, <16 x i32> <i32 11, i32 15, i32 7, i32 3, i32 10, i32 14, i32 6, i32 2, i32 9, i32 13, i32 5, i32 1, i32 8, i32 12, i32 4, i32 0>

The select will likely get merged into the preceding instruction, and then we only have to do one vrgather.vv:

    vadd.vv v20, v16, v12
    vsub.vv v20, v16, v12, v0.t
    vrgatherei16.vv v24, v20, v10

This patch bails if either of the sources are a broadcast/splat/identity shuffle, since that will usually already have some sort of cheaper lowering.

This improves performance on 525.x264_r by 4.12% with -O3 -flto -march=rva22u64_v on the spacemit-x60: https://lnt.lukelau.me/db_default/v4/nts/71?compare_to=70

### CODE REVIEW

**Comment by lukel97** [comment]
These shufflevectors come from the SLP vectorization mentioned in #119386

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 1]
```diff
@@ -5540,6 +5601,16 @@ static SDValue lowerVECTOR_SHUFFLE(SDValue Op, SelectionDAG &DAG,
     ShuffleMaskRHS.push_back(IsLHSOrUndefIndex ? -1 : (MaskIndex - NumElts));
   }
 
+  // If the mask indices are disjoint between the two sources, we can lower it
+  // as a vselect + a single source vrgather.vv. Don't do this if the operands
+  // will be splatted since they will be lowered to something cheaper like
+  // vrgather.vi anyway.
+  if (!DAG.isSplatValue(V2) && !DAG.isSplatValue(V1) &&
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:5608]: A couple comments here:
1) The other case you probably want to guard against is when either side is an identity shuffle because then the existing lowering produces a masked vrgather at worst.
2) The distinction between your code (which checks LHS and RHS masks) and the swap code just below is interesting.  The two should probably be in sync.  However, looking at your test diffs, I'm not sure the isSplatMask is actually a good idea here.

File: llvm/test/CodeGen/RISCV/rvv/fixed-vector-i8-index-cornercase.ll [hunk 2]
```diff
@@ -104,60 +104,73 @@ define <512 x i8> @two_source(<512 x i8> %a, <512 x i8> %b) {
 ; CHECK-NEXT:    addi s0, sp, 1536
 ; CHECK-NEXT:    .cfi_def_cfa s0, 0
 ; CHECK-NEXT:    andi sp, sp, -512
-; CHECK-NEXT:    vsetivli zero, 1, e8, m1, ta, ma
-; CHECK-NEXT:    vmv8r.v v24, v8
+; CHECK-NEXT:    vsetivli zero, 8, e64, m1, ta, ma
```
Comment (preames) [inline llvm/test/CodeGen/RISCV/rvv/fixed-vector-i8-index-cornercase.ll:107]: I think this particularly test case is intended to hit the generic fallback.  Can you commit a test change such that the indices aren't disjoint and then rebase?

File: llvm/test/CodeGen/RISCV/rvv/fixed-vectors-fp-shuffles.ll [hunk 3]
```diff
@@ -395,3 +395,41 @@ define <4 x half> @vrgather_shuffle_vx_v4f16_load(ptr %p) {
...
+; CHECK-NEXT:    vsetvli zero, zero, e16, m2, ta, ma
+; CHECK-NEXT:    vsext.vf2 v18, v16
+; CHECK-NEXT:    vsetvli zero, zero, e32, m4, ta, ma
+; CHECK-NEXT:    vrgatherei16.vv v8, v12, v18
+; CHECK-NEXT:    ret
+  %out = shufflevector <16 x float> %v, <16 x float> %w, <16 x i32> <i32 11, i32 15, i32 7, i32 3, i32 26, i32 30, i32 22, i32 18, i32 9, i32 13, i32 5, i32 1, i32 24, i32 28, i32 20, i32 16>
+  ret <16 x float> %out
+}
+
+define <16 x float> @shuffle_disjoint_lanes_one_splat(<16 x float> %v, <16 x float> %w) {
```
Comment (preames) [inline llvm/test/CodeGen/RISCV/rvv/fixed-vectors-fp-shuffles.ll:419]: Add an identity test please.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 4]
```diff
@@ -5197,6 +5197,67 @@ static bool isCompressMask(ArrayRef<int> Mask) {
...
+    if (Srcs[SrcIdx] == -1)
+      // Mark this source as using this lane.
+      Srcs[SrcIdx] = Src;
+    else if (Srcs[SrcIdx] != Src)
+      // The other source is using this lane: not disjoint.
+      return SDValue();
+  }
+
+  SmallVector<SDValue> SelectMaskVals;
+  for (int Lane : Srcs) {
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:5238]: Can you reverse the select here so that it uses the same order as the generic fallthrough below?  From prior experience, our vmerge.vxm matching is oddly fragile.  I'd like to remove this unrelated change if possible.

File: llvm/test/CodeGen/RISCV/rvv/fixed-vectors-fp-shuffles.ll [hunk 5]
```diff
@@ -395,3 +395,41 @@ define <4 x half> @vrgather_shuffle_vx_v4f16_load(ptr %p) {
...
+; CHECK-NEXT:    vsetivli zero, 16, e32, m4, ta, mu
+; CHECK-NEXT:    vle16.v v20, (a0)
+; CHECK-NEXT:    lui a0, 15
+; CHECK-NEXT:    addi a0, a0, 240
+; CHECK-NEXT:    vmv.s.x v0, a0
+; CHECK-NEXT:    vrgather.vi v16, v8, 7
+; CHECK-NEXT:    vrgatherei16.vv v16, v12, v20, v0.t
+; CHECK-NEXT:    vmv.v.v v8, v16
+; CHECK-NEXT:    ret
+  %out = shufflevector <16 x float> %v, <16 x float> %w, <16 x i32> <i32 7, i32 7, i32 7, i32 7, i32 26, i32 30, i32 22, i32 18, i32 7, i32 7, i32 7, i32 7, i32 24, i32 28, i32 20, i32 16>
```
Comment (preames) [inline llvm/test/CodeGen/RISCV/rvv/fixed-vectors-fp-shuffles.ll:433]: This test isn't exactly convincing that we want to avoid the splat case.  I think maybe this hints at the difference between a splatvalue (from scalar) and the splat shuffle (from mask).  The later requires a vrgather.vi, whereas the former is a vmerge.vxm.  Maybe rework the tests to cover both cases separately?

**Review by preames — CHANGES_REQUESTED** [review]
Sometimes great minds think alike.  I literally noticed this case last night, and was going to write up the patch this morning.  :)

File: llvm/test/CodeGen/RISCV/rvv/fixed-vectors-int-shuffles.ll [hunk 6]
```diff
@@ -1021,3 +1014,41 @@ define <8 x i32> @shuffle_repeat4_singlesrc_e32(<8 x i32> %v) {
...
+; CHECK-NEXT:    ret
+  %out = shufflevector <16 x i32> %v, <16 x i32> %w, <16 x i32> <i32 11, i32 15, i32 7, i32 3, i32 26, i32 30, i32 22, i32 18, i32 9, i32 13, i32 5, i32 1, i32 24, i32 28, i32 20, i32 16>
+  ret <16 x i32> %out
+}
+
+define <16 x i32> @shuffle_disjoint_lanes_one_splat(<16 x i32> %v, <16 x i32> %w) {
+; CHECK-LABEL: shuffle_disjoint_lanes_one_splat:
+; CHECK:       # %bb.0:
+; CHECK-NEXT:    lui a0, %hi(.LCPI71_0)
+; CHECK-NEXT:    addi a0, a0, %lo(.LCPI71_0)
```
[the remark this answers is not in the document]
Comment (topperc) [reply llvm/test/CodeGen/RISCV/rvv/fixed-vectors-int-shuffles.ll:1042]: It's possible to add comments from RISCVAsmPrinter.cpp. X86 does it. Unfortunately, the constant pool reference belongs to the load and not the vrgather. On X86, shuffles can often fold the load so they become the same instruction.

Let me see if I can at least print the constant pool contents on a vector load.

[the remark this answers is not in the document]
Comment (topperc) [reply llvm/test/CodeGen/RISCV/rvv/fixed-vectors-int-shuffles.ll:1042]: Hmm. Its even harder because the constant pool reference belongs to an AUIPC or ADDI instruction that is the input to the load.

File: llvm/test/CodeGen/RISCV/rvv/fixed-vectors-fp-shuffles.ll (hunk 5 above)
Comment (lukel97) [reply llvm/test/CodeGen/RISCV/rvv/fixed-vectors-fp-shuffles.ll:433]: Here's a diff from a regression if we remove the isSplatValue guard:

```diff
 ; CHECK-LABEL: shuffle_vx_v4i16:
 ; CHECK:       # %bb.0:
 ; CHECK-NEXT:    vsetivli zero, 4, e16, mf2, ta, ma
-; CHECK-NEXT:    vmv.v.i v0, 6
-; CHECK-NEXT:    vmerge.vim v8, v8, 5, v0
+; CHECK-NEXT:    vmv.v.i v0, 9
+; CHECK-NEXT:    vmv.v.i v9, 5
+; CHECK-NEXT:    vmerge.vvm v8, v9, v8, v0
 ; CHECK-NEXT:    ret
   %s = shufflevector <4 x i16> %x, <4 x i16> <i16 5, i16 5, i16 5, i16 5>, <4 x i32> <i32 0, i32 5, i32 6, i32 3>
   ret <4 x i16> %s
```

And here's one if remove the isSplatMask guard:

```diff
--- a/llvm/test/CodeGen/RISCV/rvv/fixed-vectors-int-shuffles.ll
+++ b/llvm/test/CodeGen/RISCV/rvv/fixed-vectors-int-shuffles.ll
@@ -364,12 +364,14 @@ define <8 x i8> @splat_ve4_ins_i1ve3(<8 x i8> %v) {
 define <8 x i8> @splat_ve2_we0(<8 x i8> %v, <8 x i8> %w) {
 ; CHECK-LABEL: splat_ve2_we0:
 ; CHECK:       # %bb.0:
+; CHECK-NEXT:    vsetivli zero, 8, e8, mf2, ta, ma
+; CHECK-NEXT:    vmv.v.i v0, 4
 ; CHECK-NEXT:    li a0, 66
-; CHECK-NEXT:    vsetivli zero, 8, e8, mf2, ta, mu
+; CHECK-NEXT:    vmerge.vvm v9, v9, v8, v0
 ; CHECK-NEXT:    vmv.s.x v0, a0
-; CHECK-NEXT:    vrgather.vi v10, v8, 2
-; CHECK-NEXT:    vrgather.vi v10, v9, 0, v0.t
-; CHECK-NEXT:    vmv1r.v v8, v10
+; CHECK-NEXT:    vmv.v.i v8, 2
+; CHECK-NEXT:    vmerge.vim v10, v8, 0, v0
+; CHECK-NEXT:    vrgather.vv v8, v9, v10
 ; CHECK-NEXT:    ret
   %shuff = shufflevector <8 x i8> %v, <8 x i8> %w, <8 x i32> <i32 2, i32 8, i32 2, i32 2, i32 2, i32 2, i32 8, i32 2>
   ret <8 x i8> %shuff
```

I think ideally we would really only want to do this lowering when we know we're going to end up with at least a vrgather.vv. Maybe in a follow up this could be reworked as a late combine on VRGATHER_VV_VL instead? It might be a bit trickier because you would need to reconstruct the original mask.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp (hunk 4 above)
Comment (lukel97) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp:5238]: Good point, I've reversed it to match the generic fallthrough when SwapOps is false. But it's worth pointing out that this simultaneously removes some mask diffs and introduces some others too

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp (hunk 1 above)
Comment (lukel97) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp:5607]: Added the guard for the identity shuffle.

If isSplatMask is true for either source then the original lowering would have been a vrgather.vi + vrgather.vv at worse. So I don't think there would be much benefit to replacing that with vmerge.vvm + vrgather.vv, I attached an example diff in https://github.com/llvm/llvm-project/pull/119401#discussion_r1879288787

File: llvm/test/CodeGen/RISCV/rvv/fixed-vectors-interleaved-access.ll [hunk 7]
```diff
@@ -183,406 +183,499 @@ define {<8 x i64>, <8 x i64>, <8 x i64>, <8 x i64>, <8 x i64>, <8 x i64>} @load_
 ; RV32-NEXT:    addi sp, sp, -16
 ; RV32-NEXT:    .cfi_def_cfa_offset 16
 ; RV32-NEXT:    csrr a2, vlenb
-; RV32-NEXT:    slli a3, a2, 6
```
Comment (preames) [inline llvm/test/CodeGen/RISCV/rvv/fixed-vectors-interleaved-access.ll:186]: Just for the record, this test is a real regression.  The new lower appears to increase register pressure.

I'm not worried about this case because:
1) This test case is really stressing legalization, and isn't representative of real code.
2) The two argument shuffle could have been a vcompress to start with.
3) The "right" lowering probably would have been multiple seg6 loads with offset addresses, and vslideups on the results, but that's well beyond anything worth bothering with.

File: llvm/test/CodeGen/RISCV/rvv/fixed-vectors-shuffle-deinterleave.ll [hunk 8]
```diff
@@ -11,21 +11,18 @@ define void @deinterleave3_0_i8(ptr %in, ptr %out) {
...
-; CHECK-NEXT:    vmv.s.x v9, a0
-; CHECK-NEXT:    li a0, 56
 ; CHECK-NEXT:    vmv.s.x v0, a0
-; CHECK-NEXT:    vcompress.vm v11, v8, v9
 ; CHECK-NEXT:    vsetivli zero, 8, e8, m1, ta, ma
-; CHECK-NEXT:    vslidedown.vi v8, v8, 8
-; CHECK-NEXT:    vsetivli zero, 8, e8, mf2, ta, mu
-; CHECK-NEXT:    vrgather.vv v11, v8, v10, v0.t
-; CHECK-NEXT:    vse8.v v11, (a1)
+; CHECK-NEXT:    vslidedown.vi v10, v8, 8
```
Comment (preames) [inline llvm/test/CodeGen/RISCV/rvv/fixed-vectors-shuffle-deinterleave.ll:21]: This is a fascinating accident, I hadn't considered this would catch the general deinterleave case.  I'd been exploring a targeted change specific for deinterleave type cases, and this gets it generically.  Nice!

(The new codegen is fairly neutral on m1, but if the result is > m1, the vslidedown and smaller shuffle helps a ton.)

File: llvm/test/CodeGen/RISCV/rvv/fixed-vectors-shuffle-deinterleave.ll [hunk 9]
```diff
@@ -11,21 +11,18 @@ define void @deinterleave3_0_i8(ptr %in, ptr %out) {
...
+; CHECK-NEXT:    vsetivli zero, 8, e8, mf2, ta, ma
+; CHECK-NEXT:    vmerge.vvm v8, v10, v8, v0
```
Comment (preames) [inline llvm/test/CodeGen/RISCV/rvv/fixed-vectors-shuffle-deinterleave.ll:23]: I think this vmerge should fold into the previous vslidedown.  Not sure why that didn't happen?  (Separate issue)

**Review by preames — APPROVED** [review]
LGTM w/aside comments.

File: llvm/test/CodeGen/RISCV/rvv/fixed-vectors-shuffle-deinterleave.ll (hunk 9 above)
Comment (lukel97) [reply llvm/test/CodeGen/RISCV/rvv/fixed-vectors-shuffle-deinterleave.ll:23]: Probably because the LMULs/VTs are mismatched? But in this case we might be able to get away with it because the vmerge has a smaller LMUL.

**Comment by preames** [comment]
Thought of a possible follow on extension - noting it for future reference.

In addition to the select case, we can merge the masks through other cheap (linear in LMUL) instructions.  The two that occur are vslideup.vi and vslidedown.vi, but we might be able to apply other shuffle techniques (i.e. interleave).  The vslidedown.vi subcase is potentially useful for e.g. deinterleave(N) where N is even.
